### ISSUE TITLE
[Bug] Adding Mapbox layer after DeckGL layer does not put Mapbox layer above it

### ISSUE THREAD

**Issue by joshnice** [visgl/deck.gl#9222]
### Description

Came across a scenario of when I had a DeckGL Scenegraph layer in a Mapbox overlay and then added a new Mapbox layer to the map, the Mapbox layer appeared underneath the DeckGL layer.

Normal Mapbox behaviour would place the most recently added layer to the top of the map. 

However I understand this may be intended behaviour however I could not see something in the docs for it, when looking at the beforeId property.

I created an example here - https://codepen.io/joshnice/pen/LYwOeLy?editors=1000

```
<!DOCTYPE html>
<html>

<head>
    <meta charset="utf-8">
    <title>Display a map on a webpage</title>
    <meta name="viewport" content="initial-scale=1,maximum-scale=1,user-scalable=no">
    <link href="https://api.mapbox.com/mapbox-gl-js/2.15.0/mapbox-gl.css" rel="stylesheet">
    <script src="https://api.mapbox.com/mapbox-gl-js/v2.15.0/mapbox-gl.js"></script>
    <script src="https://unpkg.com/deck.gl@latest/dist.min.js"></script>
    <style>
        body {
            margin: 0;
            padding: 0;
        }

        #map {
            position: absolute;
            top: 0;
            bottom: 0;
            width: 100%;
        }

        button {
            position: absolute;
            top: 10px;
            left: 10px;
        }
    </style>
</head>

<body>
    <div id="map"></div>
    <button id="add-more-models">Add new layer</button>
    <script>

        mapboxgl.accessToken = "pk.eyJ1Ijoiam9zaG5pY2U5OCIsImEiOiJjanlrMnYwd2IwOWMwM29vcnQ2aWIwamw2In0.RRsdQF3s2hQ6qK-7BH5cKg";

        let deckGlLayer;
        let deckOverlay;

        const createSceneGraphLayer = (beforeId) => {
            return new deck.ScenegraphLayer({
                id: "deckgl-layer-one",
                type: deck.ScenegraphLayer,
                scenegraph: "https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/master/2.0/BoxAnimated/glTF-Binary/BoxAnimated.glb",
                data: [{ coords: [0, 0] },],
                sizeScale: 10,
                getPosition: (d) => d.coords,
                getOrientation: () => [0, 0, 0],
                beforeId,
                _lighting: "pbr",
            });
        }

        const map = new mapboxgl.Map({
            container: "map",
            style: {
                version: 8,
                layers: [
                    { id: "background", type: "background", paint: { "background-color": "#cccccc" } },
                    { id: "sky", type: "sky" },
                ],
                sources: {
                    "layer-geojson": {
                        type: "geojson",
                        data: {
                            type: "FeatureCollection",
                            features: [
                                {
                                    type: "Feature",
                                    geometry: {
                                        type: "Polygon",
                                        "coordinates": [
                                            [
                                                [
                                                    -0.00041750760772174544,
                                                    0.0002822042163899141
                                                ],
                                                [
                                                    -0.00041750760772174544,
                                                    -0.00022679425441651802
                                                ],
                                                [
                                                    0.0003788494938135045,
                                                    -0.00022679425441651802
                                                ],
                                                [
                                                    0.0003788494938135045,
                                                    0.0002822042163899141
                                                ],
                                                [
                                                    -0.00041750760772174544,
                                                    0.0002822042163899141
                                                ]
                                            ]
                                        ],
                                    },
                                },
                            ],
                        },
                    },
                }
            },
            center: [0, 0],
            zoom: 20,
            pitch: 60,
            bearing: 0,
            hash: true,
            useWebGL2: true
        });

        map.on("error", (event) => {
            console.error(event.error.message);
        })

        map.once("idle", () => {
            deckGlLayer = createSceneGraphLayer(undefined);
            deckOverlay = new deck.MapboxOverlay({
                interleaved: true,
                layers: [deckGlLayer],
            });

            map.addControl(deckOverlay);
        });

        document.getElementById("add-more-models").onclick = () => {
            const layer = {
                id: "layer", type: "fill", paint: { "fill-color": "red" }, source: "layer-geojson",
            }
            map.addLayer(layer, undefined);
        }

    </script>

</body>

</html>
```

### Flavors

- [X] Script tag
- [ ] React
- [ ] Python/Jupyter notebook
- [X] MapboxOverlay
- [ ] GoogleMapsOverlay
- [ ] CartoLayer
- [ ] ArcGIS

### Expected Behavior

The most recent layer to be added to the top of the map.

### Steps to Reproduce

Shown in example above.

### Environment

- Framework version: Latest
- Browser: Edge
- OS: Windows 11

**Comment by eoudejans** [comment]
From my experience the beforeId works properly using MapboxOverlay, have you checked that the beforeId is properly set to the MapBox layer id?

**Comment by joshnice** [comment]
I agree 9 times out of 10 setting the beforeId works as expected. However in the scenario I mentioned in my first comment, if a deckgl overlay is added and the deckgl layer it is adding has a undefined beforeId and then a mapbox layer is added (where the before id is also undefined) you would expect the mapbox layer to be on top of the deckgl layer, however this is not the case.

**Comment by eoudejans** [comment]
How I solved this is to make sure to always set a beforeId, using MapBox you can defined empty dummy layers that can act as separators, for instance between background and foreground layers:
backgroundlayer: beforeId = background_layer_separator
background_layer_separator
foregroundlayer_1: beforeId = foregroundlayer_2
foregroundlayer_2: beforeId = foreground_layer_separator
foreground_layer_separator

Bottom line: make sure there always is a beforeId to attach to, this solved this issue for me.

**Comment by joshnice** [comment]
Thanks for the suggestion, I did think about doing it this way. However in the end I think I ended up just removing and readding the layers with an updated beforeId (as once I readded them I knew there was a mapbox layer above my deckgl layer in the map) which also worked.

**Comment by eoudejans** [comment]
Good to hear you found another route, I would suggest that you check that the layer you remove is not entirely recreated but merely updated using a proper beforeId, otherwise it may cause unnecessary memory allocations / requests.

**Comment by timnyborg** [comment]
I'm also running into this issue since upgrading from deck.gl 8 -> 9 and converting MapboxLayers to MapboxOverlay.  Any mapbox layers I add after an initial the deck layer (without a `beforeId`) wind up underneath.  Any subsequent deck layers defined (with a `beforeId`) work as expected.
